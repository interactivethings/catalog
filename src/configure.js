import warning from './utils/warning';
import {parsePath, addLeadingSlash} from './utils/path';
import DefaultTheme from './DefaultTheme';
import DefaultResponsiveSizes from './DefaultResponsiveSizes';
import specimens from './specimens';
import requireModuleDefault from './utils/requireModuleDefault';
import NotFound from './components/Page/NotFound';

const has = (key) => (o) => o.hasOwnProperty(key);
const hasName = has('name');
const hasTitle = has('title');
const hasSrc = has('src');
const hasPages = has('pages');
const hasComponent = has('component');

const flattenPageTree = (pageTree) => {
  return pageTree
    .reduce((pages, page) => pages.concat(page.pages ? [page, ...page.pages] : [page]), [])
    .filter((page) => page.src || page.component)
    .map((page, index) => ({...page, ...(page.hideFromMenu ? undefined : {index})}));
};

export default (config) => {
  let pageId = 0;
  const basePath = addLeadingSlash(config.basePath || '/');

  const pageReducer = (pages, page) => {
    const configStyles = config.styles || [];
    const pageStyles = page.styles || [];
    const configScripts = config.scripts || [];
    const pageScripts = page.scripts || [];

    warning(
      !hasName(page),
      'The page configuration property `name` is deprecated; use `path` instead.',
      page
    );

    warning(
      hasTitle(page),
      'The page configuration property `title` is missing.',
      page
    );

    warning(
      !hasSrc(page) || typeof page.src === 'string',
      'The page configuration property `src` must be a string.',
      page
    );

    warning(
      !hasComponent(page) || typeof requireModuleDefault(page.component) === 'function',
      'The page configuration property `component` must be a React component.',
      page
    );

    warning(
      (hasSrc(page) && !hasComponent(page) && !hasPages(page)) || (!hasSrc(page) && hasComponent(page) && !hasPages(page)) || (!hasSrc(page) && !hasComponent(page) && hasPages(page)),
      'The page configuration should (only) have one of these properties: `src`, `component` or `pages`.',
      page
    );


    return [
      ...pages,
      {
        ...page,
        id: ++pageId,
        // Currently, catalog can't be nested inside other page routes, it messes up <Link> matching. Use `basePath`
        path: page.pages ? null : parsePath(page.path || page.name, {basePath}).pathname,
        pages: page.pages ? page.pages.reduce(pageReducer, []).map((p) => ({...p, superTitle: page.title})) : null,
        styles: Array.from(new Set([...configStyles, ...pageStyles])),
        scripts: Array.from(new Set([...configScripts, ...pageScripts])),
        imports: {...config.imports, ...page.imports}
      }
    ];
  };

  const pageTree = config.pages.reduce(pageReducer, []).map((p) => ({...p, superTitle: config.title}))
    .concat({
      path: parsePath('/*', {basePath}).pathname,
      id: ++pageId,
      component: NotFound,
      title: 'Page Not Found',
      superTitle: config.title,
      scripts: [],
      styles: [],
      imports: {},
      hideFromMenu: true
    });
  const pages = flattenPageTree(pageTree);

  return {
    ...config,
    // Used to check in configureRoutes() if input is already configured
    __catalogConfig: true,
    theme: {...DefaultTheme, ...config.theme},
    responsiveSizes: config.responsiveSizes || DefaultResponsiveSizes,
    specimens: {...specimens, ...config.specimens},
    basePath,
    pages,
    pageTree
  };
};
