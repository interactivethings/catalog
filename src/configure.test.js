import configure from './configure';

test('Configuration with default theme and specimens', () => {
  expect(configure({
    title: 'Catalog',
    pages: [
      {
        path: '/',
        title: 'Overview',
        src: 'overview.md'
      }
    ]
  })).toMatchSnapshot();
});

test('Configuration with nested pages', () => {
  expect(configure({
    title: 'Catalog',
    pages: [
      {
        path: '/',
        title: 'Overview',
        pages: [
          {
            path: 'foo',
            title: 'Foo',
            src: 'foo.md'
          },
          {
            path: 'bar',
            title: 'Bar',
            src: 'bar.md'
          }
        ]
      }
    ]
  })).toMatchSnapshot();
});

test('Imports are merged on pages', () => {
  const config = configure({
    title: 'Catalog',
    imports: {Foo: 'Foo'},
    pages: [
      {
        path: '/',
        title: 'Overview',
        imports: {Bar: 'Bar'},
        src: 'overview.md'
      }
    ]
  });

  expect(config.pages[0].imports).toEqual({Foo: 'Foo', Bar: 'Bar'});
});

test('basePath support', () => {
  const config = configure({
    title: 'Catalog',
    basePath: 'catalog',
    pages: [
      {
        path: '/',
        title: 'Overview',
        src: 'overview.md'
      }
    ]
  });

  expect(config.pages[0].path).toBe('/catalog');
  // Fallback page
  expect(config.pages[1].path).toBe('/catalog/*');
});
