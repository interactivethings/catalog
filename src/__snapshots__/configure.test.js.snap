// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration with component 1`] = `
Object {
  "__catalogConfig": true,
  "basePath": "/",
  "pageTree": Array [
    Object {
      "component": [Function],
      "id": 1,
      "imports": Object {},
      "pages": null,
      "path": "/",
      "scripts": Array [],
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Overview",
    },
    Object {
      "component": [Function],
      "hideFromMenu": true,
      "id": 2,
      "imports": Object {},
      "path": "/*",
      "scripts": Array [],
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Page Not Found",
    },
  ],
  "pages": Array [
    Object {
      "component": [Function],
      "id": 1,
      "imports": Object {},
      "index": 0,
      "pages": null,
      "path": "/",
      "scripts": Array [],
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Overview",
    },
    Object {
      "component": [Function],
      "hideFromMenu": true,
      "id": 2,
      "imports": Object {},
      "path": "/*",
      "scripts": Array [],
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Page Not Found",
    },
  ],
  "responsiveSizes": Array [
    Object {
      "height": 640,
      "name": "small",
      "width": 360,
    },
    Object {
      "height": 768,
      "name": "medium",
      "width": 1024,
    },
    Object {
      "height": 900,
      "name": "large",
      "width": 1440,
    },
    Object {
      "height": 1080,
      "name": "xlarge",
      "width": 1920,
    },
  ],
  "specimens": Object {
    "audio": [Function],
    "code": [Function],
    "color": [Function],
    "color-palette": [Function],
    "download": [Function],
    "hint": [Function],
    "html": [Function],
    "image": [Function],
    "raw-code": [Function],
    "react": [Function],
    "table": [Function],
    "type": [Function],
    "video": [Function],
  },
  "theme": Object {
    "background": "#F9F9F9",
    "baseFontSize": 16,
    "bgDark": "#333333",
    "bgLight": "#F2F2F2",
    "brandColor": "#003B5C",
    "checkerboardPatternDark": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAAAAACoWZBhAAAAFklEQVQI12NQBQF2EGAghQkmwXxSmADZJQiZ2ZZ46gAAAABJRU5ErkJggg==",
    "checkerboardPatternLight": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAAAAACoWZBhAAAAF0lEQVQI12P4BAI/QICBFCaYBPNJYQIAkUZftTbC4sIAAAAASUVORK5CYII=",
    "codeColor": "#00263E",
    "codeStyles": Object {
      "function": Object {
        "color": "#FF5555",
      },
      "keyword": Object {
        "color": "#3F7397",
      },
      "punctuation": Object {
        "color": "#535353",
      },
      "script": Object {
        "color": "#3F7397",
      },
      "string": Object {
        "color": "#00263E",
      },
      "tag": Object {
        "color": "#FF5555",
      },
    },
    "fontFamily": "'Roboto', sans-serif",
    "fontHeading": "'Roboto', sans-serif",
    "fontMono": "'Source Code Pro', monospace",
    "lightColor": "#D6D6D6",
    "linkColor": "#FF5555",
    "msRatio": 1.2,
    "pageHeadingBackground": "#003B5C",
    "pageHeadingHeight": 200,
    "pageHeadingTextColor": "#fff",
    "sidebarColor": "#FFFFFF",
    "sidebarColorActive": "#D1312E",
    "sidebarColorHeading": "#003B5C",
    "sidebarColorLine": "#EBEBEB",
    "sidebarColorText": "#003B5C",
    "sidebarColorTextActive": "#FF5555",
    "textColor": "#333333",
  },
  "title": "Catalog",
}
`;

exports[`Configuration with default theme and specimens 1`] = `
Object {
  "__catalogConfig": true,
  "basePath": "/",
  "pageTree": Array [
    Object {
      "id": 1,
      "imports": Object {},
      "pages": null,
      "path": "/",
      "scripts": Array [],
      "src": "overview.md",
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Overview",
    },
    Object {
      "component": [Function],
      "hideFromMenu": true,
      "id": 2,
      "imports": Object {},
      "path": "/*",
      "scripts": Array [],
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Page Not Found",
    },
  ],
  "pages": Array [
    Object {
      "id": 1,
      "imports": Object {},
      "index": 0,
      "pages": null,
      "path": "/",
      "scripts": Array [],
      "src": "overview.md",
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Overview",
    },
    Object {
      "component": [Function],
      "hideFromMenu": true,
      "id": 2,
      "imports": Object {},
      "path": "/*",
      "scripts": Array [],
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Page Not Found",
    },
  ],
  "responsiveSizes": Array [
    Object {
      "height": 640,
      "name": "small",
      "width": 360,
    },
    Object {
      "height": 768,
      "name": "medium",
      "width": 1024,
    },
    Object {
      "height": 900,
      "name": "large",
      "width": 1440,
    },
    Object {
      "height": 1080,
      "name": "xlarge",
      "width": 1920,
    },
  ],
  "specimens": Object {
    "audio": [Function],
    "code": [Function],
    "color": [Function],
    "color-palette": [Function],
    "download": [Function],
    "hint": [Function],
    "html": [Function],
    "image": [Function],
    "raw-code": [Function],
    "react": [Function],
    "table": [Function],
    "type": [Function],
    "video": [Function],
  },
  "theme": Object {
    "background": "#F9F9F9",
    "baseFontSize": 16,
    "bgDark": "#333333",
    "bgLight": "#F2F2F2",
    "brandColor": "#003B5C",
    "checkerboardPatternDark": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAAAAACoWZBhAAAAFklEQVQI12NQBQF2EGAghQkmwXxSmADZJQiZ2ZZ46gAAAABJRU5ErkJggg==",
    "checkerboardPatternLight": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAAAAACoWZBhAAAAF0lEQVQI12P4BAI/QICBFCaYBPNJYQIAkUZftTbC4sIAAAAASUVORK5CYII=",
    "codeColor": "#00263E",
    "codeStyles": Object {
      "function": Object {
        "color": "#FF5555",
      },
      "keyword": Object {
        "color": "#3F7397",
      },
      "punctuation": Object {
        "color": "#535353",
      },
      "script": Object {
        "color": "#3F7397",
      },
      "string": Object {
        "color": "#00263E",
      },
      "tag": Object {
        "color": "#FF5555",
      },
    },
    "fontFamily": "'Roboto', sans-serif",
    "fontHeading": "'Roboto', sans-serif",
    "fontMono": "'Source Code Pro', monospace",
    "lightColor": "#D6D6D6",
    "linkColor": "#FF5555",
    "msRatio": 1.2,
    "pageHeadingBackground": "#003B5C",
    "pageHeadingHeight": 200,
    "pageHeadingTextColor": "#fff",
    "sidebarColor": "#FFFFFF",
    "sidebarColorActive": "#D1312E",
    "sidebarColorHeading": "#003B5C",
    "sidebarColorLine": "#EBEBEB",
    "sidebarColorText": "#003B5C",
    "sidebarColorTextActive": "#FF5555",
    "textColor": "#333333",
  },
  "title": "Catalog",
}
`;

exports[`Configuration with nested pages 1`] = `
Object {
  "__catalogConfig": true,
  "basePath": "/",
  "pageTree": Array [
    Object {
      "id": 1,
      "imports": Object {},
      "pages": Array [
        Object {
          "id": 2,
          "imports": Object {},
          "pages": null,
          "path": "/foo",
          "scripts": Array [],
          "src": "foo.md",
          "styles": Array [],
          "superTitle": "Overview",
          "title": "Foo",
        },
        Object {
          "id": 3,
          "imports": Object {},
          "pages": null,
          "path": "/bar",
          "scripts": Array [],
          "src": "bar.md",
          "styles": Array [],
          "superTitle": "Overview",
          "title": "Bar",
        },
      ],
      "path": null,
      "scripts": Array [],
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Overview",
    },
    Object {
      "component": [Function],
      "hideFromMenu": true,
      "id": 4,
      "imports": Object {},
      "path": "/*",
      "scripts": Array [],
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Page Not Found",
    },
  ],
  "pages": Array [
    Object {
      "id": 2,
      "imports": Object {},
      "index": 0,
      "pages": null,
      "path": "/foo",
      "scripts": Array [],
      "src": "foo.md",
      "styles": Array [],
      "superTitle": "Overview",
      "title": "Foo",
    },
    Object {
      "id": 3,
      "imports": Object {},
      "index": 1,
      "pages": null,
      "path": "/bar",
      "scripts": Array [],
      "src": "bar.md",
      "styles": Array [],
      "superTitle": "Overview",
      "title": "Bar",
    },
    Object {
      "component": [Function],
      "hideFromMenu": true,
      "id": 4,
      "imports": Object {},
      "path": "/*",
      "scripts": Array [],
      "styles": Array [],
      "superTitle": "Catalog",
      "title": "Page Not Found",
    },
  ],
  "responsiveSizes": Array [
    Object {
      "height": 640,
      "name": "small",
      "width": 360,
    },
    Object {
      "height": 768,
      "name": "medium",
      "width": 1024,
    },
    Object {
      "height": 900,
      "name": "large",
      "width": 1440,
    },
    Object {
      "height": 1080,
      "name": "xlarge",
      "width": 1920,
    },
  ],
  "specimens": Object {
    "audio": [Function],
    "code": [Function],
    "color": [Function],
    "color-palette": [Function],
    "download": [Function],
    "hint": [Function],
    "html": [Function],
    "image": [Function],
    "raw-code": [Function],
    "react": [Function],
    "table": [Function],
    "type": [Function],
    "video": [Function],
  },
  "theme": Object {
    "background": "#F9F9F9",
    "baseFontSize": 16,
    "bgDark": "#333333",
    "bgLight": "#F2F2F2",
    "brandColor": "#003B5C",
    "checkerboardPatternDark": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAAAAACoWZBhAAAAFklEQVQI12NQBQF2EGAghQkmwXxSmADZJQiZ2ZZ46gAAAABJRU5ErkJggg==",
    "checkerboardPatternLight": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAAAAACoWZBhAAAAF0lEQVQI12P4BAI/QICBFCaYBPNJYQIAkUZftTbC4sIAAAAASUVORK5CYII=",
    "codeColor": "#00263E",
    "codeStyles": Object {
      "function": Object {
        "color": "#FF5555",
      },
      "keyword": Object {
        "color": "#3F7397",
      },
      "punctuation": Object {
        "color": "#535353",
      },
      "script": Object {
        "color": "#3F7397",
      },
      "string": Object {
        "color": "#00263E",
      },
      "tag": Object {
        "color": "#FF5555",
      },
    },
    "fontFamily": "'Roboto', sans-serif",
    "fontHeading": "'Roboto', sans-serif",
    "fontMono": "'Source Code Pro', monospace",
    "lightColor": "#D6D6D6",
    "linkColor": "#FF5555",
    "msRatio": 1.2,
    "pageHeadingBackground": "#003B5C",
    "pageHeadingHeight": 200,
    "pageHeadingTextColor": "#fff",
    "sidebarColor": "#FFFFFF",
    "sidebarColorActive": "#D1312E",
    "sidebarColorHeading": "#003B5C",
    "sidebarColorLine": "#EBEBEB",
    "sidebarColorText": "#003B5C",
    "sidebarColorTextActive": "#FF5555",
    "textColor": "#333333",
  },
  "title": "Catalog",
}
`;
